### PR TITLE
nixos/slim: reintroduce

### PR DESCRIPTION
**Pull request by oxij**
# `git log`

- Revert "Merge pull request #73251 from worldofpeace/remove-slim"

  This reverts commit 4583e293fd5f4df1246b658d802a1b519bc05cd8,
  reversing changes made to d6fa5402bb7d35a7e7d0574d61a271c7e88ce079.

  The fact that SLIM is not officially actively maintained is not a good
  reason to remove it. It works, and, in some instances, like automatic
  logins, it is the only thing that actually works. See
  https://github.com/NixOS/nixpkgs/pull/46396 and
  https://github.com/NixOS/nixpkgs/pull/30890#issuecomment-419670755.

- slim: switch to slim-ng

  ... which is a fork maintained by me.

- nixos/slim: save some closure space by using the default theme when using autologin

# `nix-instantiate` environment

- Host OS: Linux 4.9, SLNOS 19.09
- Nix: nix-env (Nix) 2.2.2
- Multi-user: yes
- Sandbox: yes
- NIXPKGS_CONFIG:

```nix
{
  checkMeta = true;
  doCheckByDefault = true;
}
```

# Things done

- [X] Normal login works.
- [X] Autologin works.
- [X] Special commands work.

(Tested by applying this patchset on top of NixOS 19.09, and then testing the resulting buiild on a real hardware.)

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2003.xml
Comment (rnhmjoj) [inline nixos/doc/manual/release-notes/rl-2003.xml]: Mybe mention instead that `slim` is now your `slim-ng`

**Comment by jtojnar** [comment]
The main issue with SLIM is that it does not integrate well with modern desktops like GNOME, Plasma or Sway. In particular:

* It is missing the ability to load session files from `$XDG_DATA_DIRS`, or some other way of loading desktop files from more than a single `sessiondir`. While we are currently linking upstream session files into a single tree, `$out/share/xsessions` and `$out/share/wayland-sessions` will always be different directories (session type is determined from the directory name).
* Wayland sessions are not supported.
* Desktop environments are moving towards using [systemd for starting sessions](https://blogs.gnome.org/laney/2018/06/26/starting-sessions-with-systemd/), and without upstream, adding support would fall on us.

It is misleading when a software with limited and frozen feature set is advertised as an alternative to modern maintained DMs. In order for SLIM to get back to tree, it should be de-emphasized in the documentation, and users should be warned about its state of development and supported features.

Though, to be honest, LightDM module is not on par with SDDM and GDM at the moment either (see e.g. https://github.com/NixOS/nixpkgs/issues/56342).

File: nixos/doc/manual/configuration/x-windows.xml
Comment (jtojnar) [inline nixos/doc/manual/configuration/x-windows.xml]:
```suggestion
<xref linkend="opt-services.xserver.displayManager.gdm.enable"/> = true;
<xref linkend="opt-services.xserver.displayManager.slim.enable"/> = true; # limited integration
```

**Comment by volth** [comment]
> The main issue with SLIM is that it does not integrate well with modern desktops like GNOME, Plasma or Sway. 

But it works very well with IceWM :)

File: nixos/doc/manual/development/option-declarations.xml
Comment (jtojnar) [inline nixos/doc/manual/development/option-declarations.xml]:
```suggestion
      linkend='ex-option-declaration-eot-backend-gdm' />,
    <xref
      linkend='ex-option-declaration-eot-backend-slim' />,
```

File: nixos/doc/manual/development/option-declarations.xml
Comment (jtojnar) [inline nixos/doc/manual/development/option-declarations.xml]: Better use fully-featured DM here:

```suggestion
  <example xml:id='ex-option-declaration-eot-backend-gdm'>
```

File: nixos/doc/manual/development/option-declarations.xml
Comment (jtojnar) [inline nixos/doc/manual/development/option-declarations.xml]:
```suggestion
   <title>Extending <literal>services.xserver.displayManager.enable</literal> in the <literal>gdm</literal> module</title>
```

File: nixos/doc/manual/development/option-declarations.xml
Comment (jtojnar) [inline nixos/doc/manual/development/option-declarations.xml]:
```suggestion
  type = with types; nullOr (enum [ "gdm" ]);
```

File: nixos/lib/testing-python.nix
Comment (jtojnar) [inline nixos/lib/testing-python.nix]: IMHO, we should keep the default DM in tests.

File: nixos/lib/testing.nix
Comment (jtojnar) [inline nixos/lib/testing.nix]: Same here.

File: nixos/modules/services/x11/display-managers/default.nix
Comment (jtojnar) [inline nixos/modules/services/x11/display-managers/default.nix]:
```suggestion
# program responsible for handling X logins (such as LightDM, GDM, SDDM, or
```

File: nixos/modules/services/x11/display-managers/default.nix
Comment (jtojnar) [inline nixos/modules/services/x11/display-managers/default.nix]: Let’s keep it on the down low.

```suggestion
            "''${pkgs.lightdm}/bin/lightdm"
```

File: nixos/modules/services/x11/display-managers/default.nix
Comment (jtojnar) [inline nixos/modules/services/x11/display-managers/default.nix]: I do not think it is good idea to include it in the example, as it will conflict with the SLiM module when used.

File: pkgs/applications/display-managers/slim/default.nix
Comment (jtojnar) [inline pkgs/applications/display-managers/slim/default.nix]: Can we use a tag here?

File: pkgs/applications/display-managers/slim/default.nix
Comment (jtojnar) [inline pkgs/applications/display-managers/slim/default.nix]: Aargh, why does not the build script use [`GNUInstallDirs`](https://cmake.org/cmake/help/v3.16/module/GNUInstallDirs.html) CMake module?

File: pkgs/applications/display-managers/slim/default.nix
Comment (jtojnar) [inline pkgs/applications/display-managers/slim/default.nix]: CMake builds in parallel by default.

File: pkgs/applications/display-managers/slim/default.nix
Comment (jtojnar) [inline pkgs/applications/display-managers/slim/default.nix]: Build platform dependencies like `cmake` and `pkg-config` should go to `nativeBuildInputs`

File: pkgs/applications/display-managers/slim/default.nix
Comment (jtojnar) [inline pkgs/applications/display-managers/slim/default.nix]: Missing description.

**Comment by oxij** [comment]
@rnhmjoj

> Mybe mention instead that `slim` is now your `slim-ng`

What for? It is a drop-in replacement.

@jtojnar

> * It is missing the ability to load session files from `$XDG_DATA_DIRS`, or some other way of loading desktop files from more than a single `sessiondir`. While we are currently linking upstream session files into a single tree, `$out/share/xsessions` and `$out/share/wayland-sessions` will always be different directories (session type is determined from the directory name).

This would very easy to implement, assuming this is ever required. But since Wayland support is unlikely (because pretty much everything will need to be rewritten) I don't really see a problem.

> * Wayland sessions are not supported.

Yes. Wayland support is unlikely. Unless there exists a small library that would allow to reimplement the rendering code for both X11 and Wayland in a single breath.

(Hm. Sounds like SDL. Rewriting it for SDL sounds doable and maybe even enjoyable, so maybe. Not soon, thought, I have other plans for quite awhile.)

> * Desktop environments are moving towards using [systemd for starting sessions](https://blogs.gnome.org/laney/2018/06/26/starting-sessions-with-systemd/), and without upstream, adding support would fall on us.

Looks fairly easy to implement, thought I don't particularly like the idea. IMHO, if you want to run stuff with systemd-user you can just plug it into a *.desktop file or whatever. Why should DM even care about it? Sound like another not-invented-here-so-systemd-will-have-reinvent-this-one-too thing to me.

> It is misleading when a software with limited and frozen feature set is advertised as an alternative to modern maintained DMs. In order for SLIM to get back to tree, it should be de-emphasized in the documentation, and users should be warned about its state of development and supported features.

Sure. Will fix.

**Comment by avnik** [comment]
@jtojnar I am sure, that most of slim (or slim-ng) users don't use major DEs, and use it with simple WMs (icewm/xmonad/qtile/etc). It can be mentioned in docs as recomendation to choose other one for GNOME, etc.

**Comment by worldofpeace** [comment]
@oxij This package shouldn't even be called `slim`. It isn't slim anymore, it's a package you have taken ownership of and continued. It's likely to proliferate in use. You should even rename the slim module to `slim-ng`.

Also, what about:
* systemd services/session
* the bug in the screen locker https://github.com/iwamatsu/slim/issues/4
* issues mentioned by @jtojnar 

I'd also want to **emphasize** the importance of xdg spec compliance.

And the biggest issue is wayland, it's not supported.
It's not a requirement yet, notably sddm doesn't support it, but it should be something tracked in nixpkgs as we make moves.

So I'm seeing this being blocked on some code changes.
I'm not equipped to review C++ code, since you are making changes to slim's code I feel the need for someone with the know-how to review your work and diff the changes from the prior project.

---

I mentioned how if a community member forked and continued the project we'd be in good hands.
I'm glad this has happened.

File: nixos/tests/slim.nix
Comment (worldofpeace) [inline nixos/tests/slim.nix]: Can we port this to the python driver? https://github.com/NixOS/nixpkgs/issues/72828

File: pkgs/applications/display-managers/slim/default.nix
Comment (worldofpeace) [inline pkgs/applications/display-managers/slim/default.nix]: what has makeWrapper been used for?

File: pkgs/applications/display-managers/slim/default.nix
Comment (worldofpeace) [reply pkgs/applications/display-managers/slim/default.nix]: The patch is ugly too. The build system should work in a really native way if it's literally a "For NixOS" package. `GNUInstallDirs` would be the way to go for that.
(though, not very fond of CMake)

**Comment by worldofpeace** [comment]
Can slim-ng support `default` or soon to be [`sessionDefault`](https://github.com/NixOS/nixpkgs/pull/53843)?

File: pkgs/applications/display-managers/slim/default.nix
Comment (Mic92) [reply pkgs/applications/display-managers/slim/default.nix]: if `makeWrapper` is used it should be in `nativeBuildInputs` as well.

**Comment by rnhmjoj** [comment]
> What for? It is a drop-in replacement.

Technically It's a different software. One should know it's getting your fork and not the original slim.

> This package shouldn't even be called slim. It isn't slim anymore, it's a package you have taken ownership of and continued. It's likely to proliferate in use. You should even rename the slim module to slim-ng.

Is there a policy on this topic? For example in https://github.com/NixOS/nixpkgs/pull/74831/commits/34cc444e7bcd399830c574367ab7b9f6682cfd2f compton has been (more or less) silently replaced by a fork. Imho even if by the same author this should be mentioned in the release notes and maybe the attribute name should change as well.

**Comment by worldofpeace** [comment]
> Is there a policy on this topic? For example in 34cc444 compton has been (more or less) silently replaced by a fork. Imho even if by the same author this should be mentioned in the release notes and maybe the attribute name should change as well.

In that particular example, I don't think it should even have been aliased. An alias could be used when two identities were created for the same thing, old and new. A good example was gnome-mpv was renamed to celluloid, but it's the same code. In that example you want to help out with **migraton**, so you create a `throw` indicating that compton has been removed, but `picom` is essentially a drop in replacement. This is what should be done in slim-ng as well.

**Comment by worldofpeace** [comment]
However, sometimes in the cases like with forks, the new maintainer doesn't re-identify this software.  Things become historically confusing, and the "new" source will proliferate in distributions but it's fundamentally different. It's not as bad because it's usually 1:1 compatible, but as a packager, I'd request the maintainer to re-identify the software to avoid this.

**Comment by oxij** [comment]
>> @@ -39,7 +39,7 @@
>    can select an alternative one by picking one of the following lines:
>  <programlisting>
>  <xref linkend="opt-services.xserver.displayManager.sddm.enable"/> = true;
> -<xref linkend="opt-services.xserver.displayManager.gdm.enable"/> = true;
> +<xref linkend="opt-services.xserver.displayManager.slim.enable"/> = true;
>
> ```suggestion
> <xref linkend="opt-services.xserver.displayManager.gdm.enable"/> = true;
> <xref linkend="opt-services.xserver.displayManager.slim.enable"/> = true; # limited integration
> ```

Done.

>> @@ -146,7 +146,7 @@ xlink:href="https://nixos.org/nixpkgs/manual/#sec-package-naming">
>        />), and to extend
>     it in each backend module
>     (<xref
> -      linkend='ex-option-declaration-eot-backend-gdm' />,
> +      linkend='ex-option-declaration-eot-backend-slim' />,
>
> ```suggestion
>       linkend='ex-option-declaration-eot-backend-gdm' />,
>     <xref
>       linkend='ex-option-declaration-eot-backend-slim' />,
> ```

>> @@ -167,11 +167,11 @@ services.xserver.displayManager.enable = mkOption {
>  };</screen>
>    </example>
>  
> -  <example xml:id='ex-option-declaration-eot-backend-gdm'>
> -   <title>Extending <literal>services.xserver.displayManager.enable</literal> in the <literal>gdm</literal> module</title>
> +  <example xml:id='ex-option-declaration-eot-backend-slim'>
>
> Better use fully-featured DM here:
>
> ```suggestion
>   <example xml:id='ex-option-declaration-eot-backend-gdm'>
> ```

>> @@ -167,11 +167,11 @@ services.xserver.displayManager.enable = mkOption {
>  };</screen>
>    </example>
>  
> -  <example xml:id='ex-option-declaration-eot-backend-gdm'>
> -   <title>Extending <literal>services.xserver.displayManager.enable</literal> in the <literal>gdm</literal> module</title>
> +  <example xml:id='ex-option-declaration-eot-backend-slim'>
> +   <title>Extending <literal>services.xserver.displayManager.enable</literal> in the <literal>slim</literal> module</title>
>
> ```suggestion
>    <title>Extending <literal>services.xserver.displayManager.enable</literal> in the <literal>gdm</literal> module</title>
> ```

>>  <screen>
>  services.xserver.displayManager.enable = mkOption {
> -  type = with types; nullOr (enum [ "gdm" ]);
> +  type = with types; nullOr (enum [ "slim" ]);
>
> ```suggestion
>   type = with types; nullOr (enum [ "gdm" ]);
> ```

Basically, you want me to edit examples in `doc/manual/development/option-declarations.xml`. I think this a) should be outside of the scope of this PR, b) makes no sense, these are examples for module developers, not `configuration.nix` examples.

>> @@ -261,6 +261,7 @@ in rec {
>            inherit require;
>            virtualisation.memorySize = 1024;
>            services.xserver.enable = true;
> +          services.xserver.displayManager.slim.enable = false;
>
> IMHO, we should keep the default DM in tests.
>
>> @@ -248,6 +248,7 @@ in rec {
>            inherit require;
>            virtualisation.memorySize = 1024;
>            services.xserver.enable = true;
> +          services.xserver.displayManager.slim.enable = false;
>
> Same here.

Done.

>> @@ -1,9 +1,9 @@
>  # This module declares the options to define a *display manager*, the
> -# program responsible for handling X logins (such as LightDM, GDM, or SDDM).
> -# The display manager allows the user to select a *session
> -# type*. When the user logs in, the display manager starts the
> +# program responsible for handling X logins (such as xdm, gdb, or
>
> ```suggestion
> # program responsible for handling X logins (such as LightDM, GDM, SDDM, or
> ```

Done.

>> @@ -321,14 +321,15 @@ in
>          execCmd = mkOption {
>            type = types.str;
>            example = literalExample ''
> -            "''${pkgs.lightdm}/bin/lightdm"
> +            "''${pkgs.slim}/bin/slim"
>
> Let’s keep it on the down low.
>
> ```suggestion
>             "''${pkgs.lightdm}/bin/lightdm"
> ```
>
>>            '';
>            description = "Command to start the display manager.";
>          };
>  
>          environment = mkOption {
>            type = types.attrsOf types.unspecified;
>            default = {};
> +          example = { SLIM_CFGFILE = "/etc/slim.conf"; };
>
> I do not think it is good idea to include it in the example, as it will conflict with the SLiM module when used.

Done.

>> @@ -0,0 +1,42 @@
> +{ stdenv, fetchFromGitHub, fetchpatch, cmake, pkgconfig, xorg, libjpeg, libpng

All of those done.

**Comment by oxij** [comment]
> * the bug in the screen locker https://github.com/iwamatsu/slim/issues/4

I think this should be closed by 48d25b6969c1f36fe6b505f7eb308cb1771d9d2b of slim-ng, but at the moment I wouldn't recommend using slimlock anyway, it's very far from what a proper screen locker (like physlock) should be doing.

IMHO, the code there needs a good reorganization before those issues can be fixed in a sane manner (e.g., currently slimlock copy-pastes a bunch of code from app.cpp, uhg!).

> * issues mentioned by @jtojnar 

Those are fixed.

> Can we port this to the python driver? https://github.com/NixOS/nixpkgs/issues/72828

Out of scope of this PR.

> Can slim-ng support `default` or soon to be [`sessionDefault`](https://github.com/NixOS/nixpkgs/pull/53843)?

That needs more research. The cursory look didn't help. Hopefully, this is not blocking?

> slim -> slim-ng

I'm not sure what you want here. Renaming the nixos module seems rather strange. Even if slim-ng would diverge wildly from slim I would rather keep the same module and add an option there to select between packages, since they would do essentially the same thing.

Moreover, since, at this moment, slim-ng is just slim with less bugs I would vote against renaming the package attribute too. Why add an alias when you can skip it?

> core review

Sure, if somebody wants to do it, feel free. But, in general, the code there is rather messy so don't blame me for something I didn't do, please. With slim-ng v1.3.8 I did my best without, hopefully, breaking anything.

**Comment by worldofpeace** [comment]
> Out of scope of this PR.

Porting to the python testing driver is not actually of of scope.
Because this is reintroducing a test, all tests need to be written in python as the perl driver is due for removal and tests are actively being moved over https://github.com/NixOS/nixpkgs/issues/72828.
I have no reason to treat this test like any other test being added.

**Comment by worldofpeace** [comment]
> I'm not sure what you want here. Renaming the nixos module seems rather strange. Even if slim-ng would diverge wildly from slim I would rather keep the same module and add an option there to select between packages, since they would do essentially the same thing.
> 
> Moreover, since, at this moment, slim-ng is just slim with less bugs I would vote against renaming the package attribute too. Why add an alias when you can skip it?

Looking at some commits it's slowly diverting from slim, and already behaves in slightly different ways.
In my personal experience, I think a rename could be a good move here. 

> >  core review
> Sure, if somebody wants to do it, feel free. But, in general, the code there is rather messy so don't blame me for something I didn't do, please. With slim-ng v1.3.8 I did my best without, hopefully, breaking anything.

I will relay it to the person that any design decisions or original code is not to come back and reflect on you directly :smile:

**Comment by bb2020** [comment]
> The main issue with SLIM is that it does not integrate well with modern desktops like GNOME, Plasma or Sway.

It is amazing with XMonad.

**Comment by worldofpeace** [comment]
> > The main issue with SLIM is that it does not integrate well with modern desktops like GNOME, Plasma or Sway.
> 
> It is amazing with XMonad.

The desktops mentioned there provide wayland sessions. As X has no future this should be a great concern.

File: nixos/tests/slim.nix
Comment (jtojnar) [reply nixos/tests/slim.nix]: This should just be replacing `make-test.nix` with `make-test-python.nix` above and using:

```
    start_all()
    machine.wait_for_text("Username:")
    machine.send_chars("${user.name}\n")
    machine.wait_for_text("Password:")
    machine.send_chars("${user.password}\n")

    machine.wait_for_file("${user.home}/.Xauthority")
    machine.succeed("xauth merge ${user.home}/.Xauthority")
    machine.wait_for_window("^IceWM ")

    # Make sure SLiM doesn't create a log file
    machine.fail("test -e /var/log/slim.log")
```

**Comment by rnhmjoj** [comment]
Updates? Also, is the problem in #4300 still present?

**Comment by misuzu** [comment]
`slim` seems to work fine, but `slimlock` theme is taken neither from configuration nor from `SLIM_THEMESDIR` variable.

**Comment by bb2020** [comment]
Also `services.xserver.displayManager.setupCommands` was not available in the previous version. Maybe we can find a way to implement it now.

**Comment by misuzu** [comment]
> I think `slimlock` has always been un-configurable because lack of `slimlock.conf`.

`slimlock` from regular `slim` do take theme from `SLIM_THEMESDIR` env variable so this is kind of regression.

**Comment by bb2020** [comment]
> > I think `slimlock` has always been un-configurable because lack of `slimlock.conf`.
> 
> `slimlock` from regular `slim` do take theme from `SLIM_THEMESDIR` env variable so this is kind of regression.

However other variables like `dpms_off_timeout` can not be configured. I think `SLIM_CFGFILE` environment variable should also be exported as `slimlock` would read `slim.conf` too.

File: nixos/doc/manual/release-notes/rl-2003.xml
Comment (FRidh) [reply nixos/doc/manual/release-notes/rl-2003.xml]: Release notes should not be undone

File: nixos/tests/slim.nix
Comment (FRidh) [inline nixos/tests/slim.nix]: @oxij

File: nixos/modules/services/x11/display-managers/slim.nix
Comment (FRidh) [inline nixos/modules/services/x11/display-managers/slim.nix]: add as maintainer

**Review by FRidh — COMMENTED** [review]
Given there is apparently quite the interest, I suppose with the users taking up *active* maintainership it should be fine to get this back in. Regarding compatibility, is it possible to add a check preventing it from being used with incompatible DE's?

**Comment by oxij** [comment]
@jtojnar

> tests

Done. No idea if it works, though.

@rnhmjoj

> Also, is the problem in #4300 still present?

Hm, I use it pretty much the same way as the author there, but I see no such thing here. I'll check carefully later.

@misuzu @bb2020

> slimlock

Should be fixed now.

> Also xserver.displayManager.setupCommands was not available in the previous version. Maybe we can find a way to implement it now.

Yes, we can implement it now. Will do later.

@FRidh

> Regarding compatibility, is it possible to add a check preventing it from being used with incompatible DE's?

Sure, but I don't know which ones are incompatible, I only use xmonad.

> Release notes should not be undone

Well, what should I do then? That release note is no longer valid. Should I add another one there saying that it was re-added again or what?

@ ALL

Also renamed the attribute to `slim-ng` and added an alias on popular demand.

**Comment by misuzu** [comment]
`slimlock` themes is working now, but there are still some issues. Try doing this:
1. boot your machine with slim enabled
1. enter your username and **wrong** password
1. enter your username and right password
1. observe crash without proper restart

File: pkgs/top-level/aliases.nix
Comment (worldofpeace) [inline pkgs/top-level/aliases.nix]: This isn't correct. An alias is supposed to be a bind to something that's equivalent, and we've already addressed that it isn't anymore.

```
slim = throw "slim has been removed. Please use the replacement slim-ng or a different display-manager"
```

**Comment by bb2020** [comment]
I figured out now that `SLIM_THEMESDIR` is not used in this new version. It used to be injected with [this patch](https://github.com/NixOS/nixpkgs/blob/9d857244d9f56eee71c87add7e2d9b4e81f13f3f/pkgs/applications/display-managers/slim/runtime-paths.patch) in the previous version. This one puts `slim.conf` into `/etc`. `slimlock` should also read this file and this simplifies things a lot.

File: nixos/doc/manual/configuration/x-windows.xml
Comment (wmertens) [inline nixos/doc/manual/configuration/x-windows.xml:42]: Why remove gdm? That's what most users probably want

File: nixos/doc/manual/development/option-declarations.xml
Comment (wmertens) [inline nixos/doc/manual/development/option-declarations.xml:103]: Putting slim first makes it seem like the first one users should try, whereas slim is a very leftfield alternative

**Review by wmertens — CHANGES_REQUESTED** [review]
I agree that there is a place for slim-ng in NixOS, but IMHO it should be called slim-ng and clearly mention the differences with gdm/sddm in the description of the module

**Comment by bb2020** [comment]
Now that this PR is closed, I recommend `sddm` display manager to anyone who finds this thread later on. It is lightweight and very similar to `slim`.

**Pull request closed without merging**
